### ISSUE TITLE
Obfuscated code in your source !

### ISSUE THREAD

**Issue by TriMoon** [kazuhikoarase/qrcode-generator#38]
Please remove the obfuscated code by a publicly review-able version, we don't want or need malware in a library like this...

**Comment by TriMoon** [comment]
Until then i would not advise anyone to use your code...or at least urge them to edit the file and removing the obfuscated code before usage...

**Comment by kazuhikoarase** [comment]
It is base64 encoded character data.
I think it is review-able at createStringToBytes function.
```
  /**
   * @param unicodeData base64 string of byte array.
   * [16bit Unicode],[16bit Bytes], ...
   * @param numChars
   */
  qrcode.createStringToBytes = function(unicodeData, numChars) {

    // create conversion map.
    ...
    ...
```

**Comment by kazuhikoarase** [comment]
Yes, Unicode-SJIS conversion requires a large data table.

(2(Utf-16) + 2(SJIS)) x 7070(chars) * 1.33(b64) = 37K

So I defined it in base64.
If defined with native array or object map, it would be more large.

**Comment by TriMoon** [comment]
Still is better to have a larger map that is readable by anyone without need to decode...
Plus move the SJIS-code back into its own file, to leave the main file small...

**Comment by woodser** [comment]
@kazuhikoarase We need to be able to audit this code to ensure it doesn't contain malware.  How may we do so?

**Comment by kazuhikoarase** [comment]
You could confirm how endoded string is extracted.
line 556~575
Also you see extracted is a map named 'unicodeMap' and not a malware.

**Comment by hermanho** [comment]
I think it would be better to keep the original plain text version in github, make a build script and publish a pre-build version in NPM. Does it possible?

**Comment by kazuhikoarase** [comment]
I couldn't understand why my code was possible to be a malware. I think there are no difference between raw array and base64 string in security issue. But okay, I moved back it to separate file.

**Comment by TriMoon** [comment]
@kazuhikoarase , why are you resisting to put your SJIS-code in UN-obfuscated form on github?
Resisting to do so makes people suspicious...
In case of programming code when there is the slightest suspicion it is best to AVOID that code in its total.
So you have 2 options:
  1) Publish your SJIS-code in UN-obfuscated form on github, then everyone will be willing to use your code.
  2) Keep your SJIS-code in obfuscated form on github, then only a handful of people will use your code because they don't care about anything.

**Comment by marksteward** [comment]
It's not executed - it's effectively binary data. It's common to include binary data in a format that's suitable for the language being used.

**Comment by hermanho** [comment]
@marksteward , I agree. It is a point of view that the pre-build / minified / "binary" files should include in Github repository. There is no right way in real world. For examples:

1) No "binary" in source repository and it publish in other distribution channels (cdn, github release, npm). You cannot use a hotlink (rawgit.com ?) to include the library in webpage.
Examples:
https://github.com/facebook/react
https://github.com/jquery/jquery

2) Include "binary" in source repository and other distribution channels (cdn, github release, npm). The library can be include the library in webpage with hotlink approach.
Examples:
https://github.com/twbs/bootstrap ( https://github.com/twbs/bootstrap/tree/v4-dev/dist/js )
https://github.com/zingchart/zingtouch ( https://github.com/zingchart/zingtouch/tree/master/dist )

**Comment by TriMoon** [comment]
>  @marksteward 
>  It's common to include binary data in a format that's suitable for the language being used.

Since when is it common to include binary "data" in **sources** of javascript?
Especially when that "binary data" is being parsed (eg. executed) by the javascript engine?

**Comment by marksteward** [comment]
I think there's a misunderstanding here. Can you link to the line where it's executed?

**Comment by TriMoon** [comment]
The javascript code decodes his SJIS before running the decoded code right?
How else would that code be used by the javascript engine....
Even a code that decodes into "a = [1,2,3]" or even plain "[1,2,3]" that is then parsed in the normal code as "a=[1,2,3]" **is** being executed...

**Comment by kazuhikoarase** [comment]
It seems you have not read my code.

**Comment by marksteward** [comment]
> The javascript code decodes his SJIS before running the decoded code right?

Not quite.

[(1)](https://github.com/kazuhikoarase/qrcode-generator/blob/613014ec/js/qrcode_SJIS.js#L24) the plugin calls `createStringToBytes`, passing its base64-encoded table of mappings:

```js
qrcode.createStringToBytes('AAAAAABAAEAAgAC.../+OBUP/lgY8=', 7070)
```

[(2)](https://github.com/kazuhikoarase/qrcode-generator/blob/613014ec025e19b696b57e0804ef3aaac8354404/js/qrcode.js#L1768-L1825) `createStringToBytes` initially decodes this from base64, returning a stream that looks like this:

```js
<\0\0\0\0\1\0\1\0\2\0\2...\xff\xe3\x81\x50\xff\xe5\x81\x8f>
```

[(3)](https://github.com/kazuhikoarase/qrcode-generator/blob/613014ec/js/qrcode.js#L582-L609) It then processes the stream into a map, which relates Unicode characters to their Shift-JIS equivalents:

```js
Object {'\0': 0, '\1': 1, '\2': 2, ..., '￣': 33104, '￥': 33167}
```

[(4)](https://github.com/kazuhikoarase/qrcode-generator/blob/613014ec/js/qrcode.js#L613-L637) Finally, `createStringToBytes` returns a function that looks up Unicode codepoints in this map, and outputs the corresponding single-byte or double-byte characters. This function replaces `qrcode.stringToBytes`.

At no point is the data executed - mathematical operations are performed to convert from the encoded format into the keys and values in the mapping.

I agree with @hermanho that it might be nice to include the source object and script used to generate the base64 data, but this is of limited use as this code will not change often.

Your approach in this ticket and #34 is the sort that can put people off open source software. You're relying on the good will of people usually working in their free time - it's not helpful to say "we don't want or need malware" and "you have 3 options".

**Comment by kazuhikoarase** [comment]
>  I agree with @hermanho that it might be nice to include the source object and script used to generate the base64 data, but this is of limited use as this code will not change often.

I agree too. I had written a short code and pushed it. (But no one need to generate again.)

https://github.com/kazuhikoarase/qrcode-generator/tree/master/js/misc

**Comment by TriMoon** [comment]
> @marksteward 
> Your approach in this ticket and #34 is the sort that can put people off open source software. You're relying on the good will of people usually working in their free time - it's not helpful to say "we don't want or need malware" and "you have 3 options".

It doesn't matter because we have to rely on word of mouth and "trust" explanations instead of actually being able to confirm whats going on.
I still stand by my point that obfuscated code has no place on github, be it base64 encoded "unicode" or "zipped data/code"
I and others objecting to things on "github and a like" are also people working in their free time...
If a programmer is unable to provide transparent code and gets discouraged by comments like mine then that is actually for the better of the community...

**Comment by woodser** [comment]
Reading through the comments, I'm trying to understand if the code is verifiable and has been verified to not contain malware.  Is it?  I'd like to use this in [my own project](https://github.com/cryptostorage/cryptostorage.com), but I cannot if the code is not 100% verifiable because it will be handling user funds.  Thanks in advance.

**Comment by woodser** [comment]
Just following up.  I think it is fine if the source code contains base64 or other encoded data as long as the steps to create the data can be verified and/or it can be verified that the data is not executable.  @marksteward listed steps of how the base64 encoded data is processed (thank you!).  Can we similarly get instructions to recreate the encoded data starting from an unencoded version?  I ask only because people need to be able to audit the code that they run, especially in a financial application like mine.  Any verifiable assurance that the code cannot be malicious is appreciated.  Thank you!

**Comment by kazuhikoarase** [comment]
I added this one. Would this help you?
https://github.com/kazuhikoarase/qrcode-generator/tree/master/js/misc

**Comment by franklinyu** [comment]
> It doesn't matter because we have to rely on word of mouth and "trust" explanations instead of actually being able to confirm whats going on.

@TriMoon No, you don’t have to rely on word of mouth. Read the source code about how the data is used, like what @marksteward did. I don’t think that meets the definition of “execution”.

In addition, before you know for sure that it is executed, it is misleading to call it “obfuscated code”. It should at most be called “suspicious data” (that *may* contain anything, including obfuscated code). Similarly you won’t call a suspect “murderer” on a court. Please bring down emotion as much as possible.
